src/main/java/org/dependencytrack/util/NotificationUtil.java

```diff
@@ -104,6 +106,29 @@ public static void analyzeNotificationCriteria(QueryManager qm, Vulnerability vu
         }
     }
 
+    public static void analyzeNotificationCriteria(QueryManager qm, Vulnerability vulnerability, VulnerabilityUpdateDiff vulnerabilityUpdateDiff) {
+        final Map<Long,Project> affectedProjects = new HashMap<>();
+        List<Component> components = vulnerability.getComponents();
+        if (components != null) {
+            int i = 0;
+            for (final Component c : components) {
+                i++;
+                if (!affectedProjects.containsKey(c.getProject().getId())) {
+                    affectedProjects.put(c.getProject().getId(), qm.detach(Project.class, c.getProject().getId()));
+                }
```

Comment:
This computation needs to happen outside of the `for` loop that dispatches notifications.

Otherwise `affectedProjects` will grow with each iteration, resulting in all components other than the very last one to not get the complete set of projects.